design/secret_rotation_design.md

```diff
@@ -0,0 +1,951 @@
+one secret variable (i.e. the first secret variable that it finds to be in
+error). Therefore, for bulk secret retrieval operations, there is no way of
+knowing if any other secret values are empty or for which access has been
+revoked.
+
+So unless we were to add code to Secrets Provider to follow up any errored
+bulk secret retrieval with multiple retrievals of individual secret values,
+any cleanup of Kubernetes Secrets and/or secret files must be an
+"all or nothing" operation. This means that if any secrets have been
+deleted or access to any secrets has been revoked, the Secrets Provider
+should:
+
+- For `k8s-secret` destination mode: Sanitize ALL Kubernetes Secrets
+  by removing all secret values that have been written to those Secrets.
+- For `file` destination mode: Delete ALL secret files from the secret
+  shared volume.
+- Exit with error. The Secrets Provider treats these errors as fatal
+  whether or not this occurs during initial startup or as part of periodic
+  secrets refresh.
+
+#### Detection of Content Changes
+
+There are several components that are used by the Secrets Provider in
+rendering Kubernetes Secret or secret file content. Changes to any of these
+components may result in changes to the rendered content. For example,
+rendered Kubernetes Secret and secret file content should be updated for:
+
+- Changes to Conjur secrets
+- Changes to a custom template that is configured via a ConfigMap (`file` destination mode)
+- Changes to the data field of a source Kubernetes Secret (`k8s-secret` destination mode)
+
+Rather than look for changes in all of these components separately, the
+Secrets Provider will simply look for changes in the rendered objects
+(i.e. the "finished product").
+
+For added security, the Secrets Provider will not cache raw
+Kubernetes Secret or secret file content to perform this comparison; rather,
+it will store SHA-256 checksums of the previously written content, and
+will detect changes by comparing the previous checksums with checksums
+calculated on the current rendered content.
+
+For `file` destination mode, this detection will require that the template
+for each secret group gets temporarily rendered (via `template.Execute()`)
+into a buffer so that the SHA-256 checksum can be calculated. If rendered
+content has changed, the contents of that buffer will be written
+atomically to a file in a shared volume, as described in the next section.
+
+#### Atomic File Writes for Push-to-File Mode
+
+When the Secrets Provider is running as a sidecar container in `file`
+destination mode, it renders and writes secret files asynchronously with
+respect to the application container reading those files. In order to
+guarantee that the application container reads a consistent instance of
+file content, the Secrets Provider needs to write each secret file as an
+atomic operation.
+
+This atomicity can be achieved by using the following third-party package:
+https://github.com/natefinch/atomic.
```

Comment:
My understanding is that this shouldn't pose a problem. Before replacing the file, atomic copies the permissions from the existing destination file ([code here](https://github.com/natefinch/atomic/blob/master/atomic.go#L63-L67)). We can still create that file with the permissions we want, and then pass the path to the atomic writer. Only the temp file will have 0600 permissions.
I imagine we don't need to be concerned that the newly created file will be deleted before we have a chance to write to it (which would cause the destination file to be recreated with 0600 permissions). Is that a safe assumption?

Reply:
There's nothing we do that would lead to the file ending up with 0600 permissions, right?  If the file did get deleted before it's written to, the only result would be the file would be less accessible than if we'd created it with world readable permissions. Probably safe - I was wondering more if it would end up inaccessible to the end application than if a security hole would be present.